### PR TITLE
[WIP] update to 0.9 consumer

### PR DESCRIPTION
**Pull request by talevy**

### CODE REVIEW

File: lib/logstash/inputs/kafka.rb [hunk 1]
```diff
@@ -20,168 +21,89 @@
...
-  # maximum message size the server allows or else it is possible for the producer to send
-  # messages larger than the consumer can fetch.
-  config :fetch_message_max_bytes, :validate => :number, :default => 1048576
-  # The serializer class for messages. The default decoder takes a byte[] and returns the same byte[]
-  config :decoder_class, :validate => :string, :default => 'kafka.serializer.DefaultDecoder'
-  # The serializer class for keys (defaults to the same default as for messages)
-  config :key_decoder_class, :validate => :string, :default => 'kafka.serializer.DefaultDecoder'
+  default :codec, 'plain'
 
+  config :auto_commit_interval_ms, :validate => :string, :default => "10"
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: Need docs for all these config changes

File: lib/logstash/inputs/kafka.rb [hunk 2]
```diff
@@ -20,168 +21,89 @@
...
+  config :auto_commit_reset, :validate => :string, :default => "true"
+  config :bootstrap_servers, :validate => :string, :default => "localhost:9092"
+  config :check_crcs, :validate => :string
+  config :client_id, :validate => :string, :default => "logstash"
+  config :connections_max_idle_ms, :validate => :string
+  config :enable_auto_commit, :validate => :string, :default => "true"
+  config :fetch_max_wait_ms, :validate => :string
+  config :fetch_min_bytes, :validate => :string
+  config :group_id, :validate => :string, :default => "ll"
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: `logstash` instead of `ll`?

File: lib/logstash/inputs/kafka.rb [hunk 3]
```diff
@@ -20,168 +21,89 @@
...
-    end
-    @kafka_client_queue = SizedQueue.new(@queue_size)
-    @consumer_group = create_consumer_group(options)
-    @logger.info('Registering kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
+    @message_count = 0
+    @runner_threads = []
   end # def register
 
   public
   def run(logstash_queue)
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: Now that we've removed `kafka_client_queue`, just call this `queue` to be consistent with other plugins? wdyt?

File: logstash-input-kafka.gemspec [hunk 4]
```diff
@@ -1,7 +1,7 @@
 Gem::Specification.new do |s|
 
   s.name            = 'logstash-input-kafka'
-  s.version         = '2.0.2'
+  s.version         = '2.1.0'
```
Comment (suyograo) [inline logstash-input-kafka.gemspec]: Oh, this needs to be 3.0.0. It breaks compatibility. We are removing 90% of the old configuration :)

Comment (talevy) [inline logstash-input-kafka.gemspec]: agreed

File: lib/logstash/inputs/kafka.rb (hunk 1 above)
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: I would just copy it from http://people.apache.org/~nehanarkhede/kafka-0.9-producer-javadoc/doc/org/apache/kafka/clients/consumer/ConsumerConfig.html

File: lib/logstash/inputs/kafka.rb [hunk 5]
```diff
@@ -20,168 +21,89 @@
...
-  def create_consumer_group(options)
-    Kafka::Group.new(options)
+  def thread_runner(logstash_queue, consumer)
+    Thread.new do
+      begin
+        consumer.subscribe(topics);
+        while !stop?
+          records = consumer.poll(timeout_ms);
+          for record in records do
+            break if !max_messages.nil? && @message_count >= max_messages
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: hmm, not sure if we need this. LS's current design is to stream forever, and wait until new messages are available.

File: lib/logstash/inputs/kafka.rb [hunk 6]
```diff
@@ -20,168 +21,89 @@
...
-                            'key' => message_and_metadata.key}
-        end
-        output_queue << event
-      end # @codec.decode
-    rescue => e # parse or event creation error
-      @logger.error('Failed to create event', :message => "#{message_and_metadata.message}", :exception => e,
-                    :backtrace => e.backtrace)
-    end # begin
-  end # def queue_event
+  def new_consumer
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: `create_consumer` is more meaningful

File: lib/logstash/inputs/kafka.rb [hunk 7]
```diff
@@ -20,168 +21,89 @@
...
-      end
-
-      @logger.info('Done running kafka input')
-    rescue => e
-      @logger.warn('kafka client threw exception, restarting',
-                   :exception => e)
-      Stud.stoppable_sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000) { stop? }
-      retry if !stop?
-    end
+    @runner_consumers = num_threads.times.map { || new_consumer }
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: From Kafka's documentation - http://people.apache.org/~nehanarkhede/kafka-0.9-consumer-javadoc/doc/ 

> The consumer is thread safe and should generally be shared among all threads for best performance.

Looks like we just need one consumer instance and share it across threads...

File: lib/logstash/inputs/kafka.rb [hunk 8]
```diff
@@ -20,168 +21,89 @@
...
-    Kafka::Group.new(options)
+  def thread_runner(logstash_queue, consumer)
+    Thread.new do
+      begin
+        consumer.subscribe(topics);
+        while !stop?
+          records = consumer.poll(timeout_ms);
+          for record in records do
+            break if !max_messages.nil? && @message_count >= max_messages
+            @message_count = @message_count.next
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: I'd remove this feature unless someone explicitly asks for consuming `n` events and shutting down. WDYT?

File: lib/logstash/inputs/kafka.rb (hunk 7 above)
Comment (talevy) [inline lib/logstash/inputs/kafka.rb]: For whatever reason, those snapshot docs seem to have higher SEO than the real release docs.

as far as the official docs say here: http://kafka.apache.org/090/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html

> The consumer maintains TCP connections to the necessary brokers to fetch data. Failure to close the consumer after use will leak these connections. The consumer is not thread-safe. See [Multi-threaded Processing](http://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#multithreaded) for more details.

File: lib/logstash/inputs/kafka.rb (hunk 6 above)
Comment (talevy) [inline lib/logstash/inputs/kafka.rb]: oh naming :) I had that written at one point, was flipflopping between the two. I will switch to `create_consumer`

File: Rakefile [hunk 9]
```diff
@@ -1,7 +1,28 @@
...
+desc "Get jars"
+task :vendor do
+  jar_target = "vendor/jar-dependencies/runtime-jars"
+  kafka_version = "0.9.0.0"
+  kafka_url = "http://central.maven.org/maven2/org/apache/kafka/kafka-clients/#{kafka_version}/kafka-clients-#{kafka_version}.jar"
+  slf4j_version = "1.7.13"
+  slf4j_url = "http://central.maven.org/maven2/org/slf4j/slf4j-api/#{slf4j_version}/slf4j-api-#{slf4j_version}.jar"
+  slf4j_nop_url = "http://central.maven.org/maven2/org/slf4j/slf4j-nop/#{slf4j_version}/slf4j-nop-#{slf4j_version}.jar"
+
+  puts "Will get jars for version #{kafka_version}"
```
Comment (suyograo) [inline Rakefile]: nitpick, but :)

"Downloading jars for version #{kafka_version}"

File: Rakefile [hunk 10]
```diff
@@ -1,7 +1,28 @@
...
+  puts "Removing current jars"
+  `rm -rf #{jar_target}`
+  `mkdir -p #{jar_target}`
+  Dir.chdir jar_target
+  puts "Will download #{kafka_url}"
```
Comment (suyograo) [inline Rakefile]: Downloading

File: Rakefile [hunk 11]
```diff
@@ -1,7 +1,28 @@
...
+  `curl #{kafka_url} -o kafka-clients-#{kafka_version}.jar`
+  puts "Will download #{slf4j_url}"
+  `curl #{slf4j_url} -o slf4j-api-#{slf4j_version}.jar`
+  puts "Will download #{slf4j_nop_url}"
```
Comment (suyograo) [inline Rakefile]: Downloading

File: lib/logstash/inputs/kafka.rb (hunk 7 above)
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: Ugh, thanks for pointing that out! I see the pros/cons they describe in http://kafka.apache.org/090/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html

Your approach looks good, based on their example

File: lib/logstash/inputs/kafka.rb [hunk 12]
```diff
@@ -20,168 +21,89 @@
...
+  def new_consumer 
+    props = java.util.Properties.new
+    props.put("bootstrap.servers", bootstrap_servers)
+    props.put("group.id", group_id);
+    props.put("enable.auto.commit", enable_auto_commit);
+    props.put("auto.commit.interval.ms", auto_commit_interval_ms);
+    props.put("session.timeout.ms", session_timeout_ms);
+    props.put("key.deserializer", key_deserializer);
+    props.put("value.deserializer", value_deserializer);
+    org.apache.kafka.clients.consumer.KafkaConsumer.new(props);
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: We also need to pass all the configs here, does not look like the full list we expose above

File: lib/logstash/inputs/kafka.rb [hunk 13]
```diff
@@ -20,168 +21,89 @@
...
-  # request. These bytes will be read into memory for each partition, so this helps control
-  # the memory used by the consumer. The fetch request size must be at least as large as the
-  # maximum message size the server allows or else it is possible for the producer to send
-  # messages larger than the consumer can fetch.
-  config :fetch_message_max_bytes, :validate => :number, :default => 1048576
-  # The serializer class for messages. The default decoder takes a byte[] and returns the same byte[]
-  config :decoder_class, :validate => :string, :default => 'kafka.serializer.DefaultDecoder'
-  # The serializer class for keys (defaults to the same default as for messages)
-  config :key_decoder_class, :validate => :string, :default => 'kafka.serializer.DefaultDecoder'
+  default :codec, 'plain'
```
Comment (suyograo) [inline lib/logstash/inputs/kafka.rb]: so, we are changing the codec from `json` to `plain`?

**Comment by suyograo** [comment]
@talevy tested this consumer on 0.9.0 broker, works good. I was not able to run 0.8.2  broker after I deployed 0.9.0 broker (it maybe a local metadata issue).. I'd like to check if this is indeed backward compatible..

**Comment by joekiller** [comment]
You should be able to run 0.8 producers and consumers with a 0.9 broker.

**Comment by suyograo** [comment]
Thanks @joekiller. I was referring to the other way around - running 0.9 consumer with 0.8 broker as the docs mention it is backward compatible. Wanted to test it explicitly.

**Comment by talevy** [comment]
closing in favor of continuing development in this branch: https://github.com/logstash-plugins/logstash-input-kafka/tree/update/0.9

**Pull request closed without merging**